src/lib/falcor/falcorUtils.js

```diff
@@ -40,95 +40,110 @@ export function validateFalcorPathSets(falcorPathSets) {
+    }
+
+    return undefined;
+  }, object);
+}
+
+export function pathSetsInCache(cache, falcorPathSets) {
+  function handleCheckingSingleKey(curObject, nextRemainingKeySets, key) {
+    if (!curObject.hasOwnProperty(key)) {
+      return false;
+    }
+    else {
+      const val = curObject[key];
+      if (val.$type) {
+        switch (val.$type) {
+          case "error":
+          case "atom":
+            return nextRemainingKeySets.length === 0;
+          case "ref":
+            return checkSinglePathSetInCache(followPath(val.value, cache), nextRemainingKeySets);
+          default:
+            throw new Error("pathSetsInCache encountered unexpected type. Type found was: " + val.$type);
         }
-        end = key.to;
-      }
-      else if (key.hasOwnProperty("length")) {
-        end = start+key.length-1;
       }
       else {
-        throw new Error("Falcor Range must have either 'to' or 'length' properties at falcor KeySet: " + JSON.stringify(key));
+        return checkSinglePathSetInCache(val, nextRemainingKeySets);
       }
-      for (let i = start; i <= end; i++) {
-        if (curObject.hasOwnProperty(i)) {
-          if (!recurseOnCache(curObject[i], nextRemainingKeySets)) {
-            return false;
+    }
+  }
+
+  function checkSinglePathSetInCache(curObject, remainingKeySets) {
+    if (remainingKeySets.length === 0) {
+      return true;
+    }
+    const nextRemainingKeySets = remainingKeySets.slice(1);
+    let nextKeySet = remainingKeySets[0];
+    // This is to avoid code duplication for when
+    // it is just a single key instead of an
+    // array of keys. We don't want to handle that
+    // case seperately so just make an array with one value
+    if (!(nextKeySet instanceof Array)) {
+      nextKeySet = [nextKeySet];
+    }
+    return nextKeySet.every((keyOrRange) => {
+      if (keyOrRange !== null && typeof keyOrRange === "object") {
+        // keyOrRange is a range
+        let start = 0;
+        if (keyOrRange.hasOwnProperty("from")) {
+          start = keyOrRange.from;
+        }
+        let end;
+        if (keyOrRange.hasOwnProperty("to")) {
+          if (keyOrRange.hasOwnProperty("length")) {
+            throw new Error("Falcor Range cannot have both 'to' and 'length' properties at falcor KeySet: " + JSON.stringify(keyOrRange));
           }
+          end = keyOrRange.to;
+        }
+        else if (keyOrRange.hasOwnProperty("length")) {
+          end = start+keyOrRange.length-1;
         }
         else {
-          return false;
+          throw new Error("Falcor Range must have either 'to' or 'length' properties at falcor KeySet: " + JSON.stringify(keyOrRange));
         }
-      }
-    }
-    else {
-      if (curObject.hasOwnProperty(key)) {
-        if (!recurseOnCache(curObject[key], nextRemainingKeySets)) {
-          return false;
+        for (let i = start; i <= end; i++) {
+          if (!handleCheckingSingleKey(curObject, nextRemainingKeySets, i)) {
```

Comment:
Yes that's actually very true. I guess the simple solve would just to let it return false and do a normal fetch, but since we'll be overfetching quite often we'd of course love to have something better than that. 

I guess one possibility could be that every time we handle a range (because we would of course never overfetch with anything but a range) we just check if there is a single key in the range that exists in the cache, then we say the whole range is there. If we were really cheap about it we could just return true if the first item in the range was there which would also greatly improve performance of the check. 
If we wanted to be more thorough, and check that it is an overfetch, and randomly missing data, one check could be that the only existing keys are allowed to be consecutive keys from the start of the range we are fetching, and as soon as we find a "hole" in the range with no data, then it should be empty until the end of the range.

I think that would work quite well, either of the two options above I just listed. I guess in theory there could be weird edge cases that could break it, but I don't feel like that would happen in practice. What do you say?